File: lang/src/main/scala/uk/co/zentium/langscala/Memoization.scala
package uk.co.zentium.langscala

import org.scalatest.Matchers

import scala.collection.mutable

object Memoization extends App with Matchers {

  val checkPrime: Int => Boolean = {

    def checkPrimeInner(x: Int) = 2 until x forall (v => x % v != 0)

    val cache = mutable.Map[Int, Boolean]()
    x => {
      val now: Long = System.nanoTime()
      if (cache.contains(x)) {
        println("from cache took " + (System.nanoTime() - now))
      } else {
        cache(x) = checkPrimeInner(x)
        println("not from cache took " + (System.nanoTime() - now))
      }
      cache(x)
    }
  }

  println(checkPrime(101))
  println(checkPrime(101))
  println(checkPrime(500))
  println(checkPrime(500))

}

File: lang/src/main/java/uk/co/zentium/langjava/Memoization.java
package uk.co.zentium.langjava;

import java.util.HashMap;
import java.util.Map;

public class Memoization {

    private static Map<Integer, Boolean> cache = new HashMap<>();

    private boolean checkPrimeInner(int x) {
        boolean isPrime = true;
        for (int i = 2; i < x; i++) {
            if (x % i == 0) {
                isPrime = false;
            }
        }
        return isPrime;
    }

    boolean checkPrime(int x) {
        long now = System.nanoTime();
        if (cache.containsKey(x)) {
            System.out.println("from cache took " + (System.nanoTime() - now));
        } else {
            cache.put(x, checkPrimeInner(x));
            System.out.println("not from cache took " + (System.nanoTime() - now));
        }
        return cache.get(x);
    }

    public static void main(String[] args) {
        Memoization memoization = new Memoization();
        System.out.println(memoization.checkPrime(101));
        System.out.println(memoization.checkPrime(101));
        System.out.println(memoization.checkPrime(500));
        System.out.println(memoization.checkPrime(500));
    }
}